gh-60691: allow certificates to be specified from memory

jgehrcke opened the pull request:
Refs:
- http://bugs.python.org/issue16487
- http://bugs.python.org/review/16487

Compared to the state last discussed via http://bugs.python.org/review/16487, I have
- rebased on latest master and resolved merge conflicts (nothing complicated)
- incorporated the changes that I had acked in [this review board message](http://bugs.python.org/review/16487/#msg3)


<!-- issue-number: bpo-16487 -->
https://bugs.python.org/issue16487
<!-- /issue-number -->


<!-- gh-issue-number: gh-60691 -->
* Issue: gh-60691
<!-- /gh-issue-number -->

jgehrcke committed:
bpo-16487: allow certificates to be specified from memory

pitrou commented on Lib/test/test_ssl.py:
```diff
@@ -123,6 +124,24 @@ def asn1time(cert_time):
+    """
+    with open(filepath, 'rb') as f:
+        buf = BytesIO(f.read())
+    buf.read = buf.getvalue
```
Why do you need that? It should work without.
```
>>> b = io.BytesIO(b"abc")
>>> b.read()
b'abc'
```

jgehrcke replied:
I think I wanted to have the object be re-`read()`able, because IIRC that simplified writing the test code quite a bit (it probably allowed me to prepare the various file objects just once, instead of for every `load_cert_chain()` method invocation):

```
>>> import io
>>> b = io.BytesIO(b"abc")
>>> b.read()
b'abc'
>>> b.read()
b''
```
vs.
```
>>> b = io.BytesIO(b"abc")
>>> b.read = b.getvalue
>>> b.read()
b'abc'
>>> b.read()
b'abc'
```

pitrou commented on Modules/_ssl.c:
```diff
@@ -3289,6 +3291,200 @@ _ssl__SSLContext_load_cert_chain_impl(PySSLContext *self, PyObject *certfile,
+    }
+
+
+    PySSL_BEGIN_ALLOW_THREADS_S(pw_info.thread_state);
```
Can a password be asked for the certificates (as opposed to the private key)? This sounds a bit weird.

jgehrcke replied:
The "certificate file" is actually allowed to contain the private key: https://docs.python.org/3/library/ssl.html#combined-key-and-certificate. That's one of the reasons why this function's input space is so gigantic.

In an extreme (but sane) case this file contains a password-protected private key, an end-entity certificate, and N intermediate CA certificates.

Edit: That argument might not be satisfying, I had all the relevant code comments in place in my dev branch a year ago -- will try to clarify that in a future comment here. In any case this behavior is tested here: https://github.com/python/cpython/pull/2449/files#diff-e144a9cacd10921d4dae0aeac0300a6fR1045

jgehrcke replied:
Hey @pitrou. I got around to looking into this again. My argument above was not the correct argument to answer your question. I went into the code and docs again and understood the reasoning again. There are two points of relevance here:

**1) cert files might indeed be encrypted**

From https://www.openssl.org/docs/man1.0.2/ssl/SSL_CTX_use_certificate_chain_file.html:

> The private keys loaded from file can be encrypted. In order to successfully load encrypted keys, a function returning the passphrase must have been supplied, see SSL_CTX_set_default_passwd_cb. (**Certificate files might be encrypted as well** from the technical point of view, it however does not make sense as the data in the certificate is considered public anyway.)

(emphasis mine)

**2) Consistency**

That is the code block in question from my patch:

```
    PySSL_BEGIN_ALLOW_THREADS_S(pw_info.thread_state);
    r = _openssl_use_certificate_chain_from_bio(self->ctx, certbio->bio);
    PySSL_END_ALLOW_THREADS_S(pw_info.thread_state);
```

whereas ` _openssl_use_certificate_chain_from_bio()` is by definition meant to _closely_ resemble OpenSSL's `SSL_CTX_use_certificate_chain_file()`.  Both *can* handle the password callback. As of reason (1), I suppose. From the [source](https://github.com/openssl/openssl/blob/d72a00416a0691bfd4920008767221bb4082a2ed/ssl/ssl_rsa.c#L589) of `SSL_CTX_use_certificate_chain_file()` in OpenSSL 1.1.0:
```
    x = PEM_read_bio_X509_AUX(in, NULL, passwd_callback,
                              passwd_callback_userdata);
```

 *If* `_openssl_use_certificate_chain_from_bio()` or `SSL_CTX_use_certificate_chain_file()` triggers the password callback I believe we better have the wrapper in place.

That's probably what one of the earlier authors thought, too, and so the same ALLOW_THREADS wrapper around `SSL_CTX_use_certificate_chain_file()` has always been [present in CPython so far](https://github.com/python/cpython/blob/40c54d5e1aaab91cb7df71f735112d20b5e5b755/Modules/_ssl.c#L3323):
```
    PySSL_BEGIN_ALLOW_THREADS_S(pw_info.thread_state);
    r = SSL_CTX_use_certificate_chain_file(self->ctx,
        PyBytes_AS_STRING(certfile_bytes));
    PySSL_END_ALLOW_THREADS_S(pw_info.thread_state);
```

Notably, I do not think that we cover the case of loading an encrypted certificate (where there is no key also present in the cert file) within `test_ssl`.

jgehrcke replied:
I have added a commit with a hopefully clarifying code comment.

pitrou commented on Lib/ssl.py:
```diff
@@ -450,6 +451,51 @@ def load_default_certs(self, purpose=Purpose.SERVER_AUTH):
+               keyfile_path = keyfile
+
+           # Pre CPython 3.7 behavior: let OpenSSL consume the files via
+           # SSL_CTX_use_certificate_chain_file().
```
Why not read the files in Python and fall through below? This would allow deleting a bunch of C code.

jgehrcke replied:
It's intriguing. You are asking this now and I feel like "mhm, I think there was a pretty strong reason to not do it, because I tried hard, but what was the reason again?".

I am glad that I wrote [a rather complete summary](http://bugs.python.org/msg272585) one year ago. It revived my memories... this is an excerpt from it:

```

[...]

An important note on backwards compatibility:

>> Antoine, are you suggesting that we remove the current c-level
>> capability to use file system files (using open()) and just go with
>> raw bytes data at the C api level, and then do the 'filename or
>> filelike object' in Python land?
>
> Yes, I think that's reasonable. Hopefully it won't break any existing uses.

I absolutely agree that Python code should be responsible for doing the
distinction between file paths and file objects. However, I do not agree with

> remove the current c-level capability to use file system files

After following the code I am rather sure that we should leave the current C
implementation (`_ssl__SSLContext_load_cert_chain_impl()` in Modules/_ssl.c)
as untouched as possible, and not try to integrate the new logic into that
very same function.

[...]

The central argument for having two separate C code paths is that old code
must proceed using OpenSSL's SSL_CTX_use_certificate_chain_file().

[...]
```

A bit of background on this can be found in the summary I linked to. I tried hard to understand the OpenSSL docs and the corresponding C code in the CPython repository, and tried hard to come up with what you were suggesting but I failed achieving the level of confidence that's required here. I was unable to achieve the desired level of confidence as of inherently bad OpenSSL API docs, and (this is what I vaguely remember) a few obvious mismatches between code and docs. In my summary I also wrote

```
When inspecting the OpenSSL code for SSL_CTX_use_certificate_chain_file(), some
of the above statements are not entirely correct.
```

(referring to statements from API docs)

So I could not rely on doc. Then I probably thought that the only way to gain confidence was via a proper test suite. In this particular case here, however, we probably agree that a test suite that actually matters does not really exist. It's comprised of all the applications in the world that use this code path. As of certificate diversity and pure combinatorics the input space into this function is super large.

From here it was rather easy to conclude for me that the old code path is better left intact, and that having a switch/shunt at a higher level is much safer in terms of backwards compatibility.

jgehrcke committed:
Merge remote-tracking branch 'origin/master' into address-issue-16487

Manually resolved conflicts in

	Lib/test/test_ssl.py
	Modules/clinic/_ssl.c.h

jgehrcke committed:
_ssl.c: address 'dereferencing pointer to incomplete type' errors

jgehrcke committed:
test_ssl: make it pass on OpenSSL 1.1.0

jgehrcke committed:
_ssl.c: improve comments

jgehrcke commented:
@pitrou @orsenthil @tiran I would very much appreciate to get another round of feedback. I am available for pushing this further, so that maybe we can still land this in 3.7. With all the information present across the issue tracker and this GH conv I want to stress again what's probably easy to miss: this is meant to be a conceptually backwards-compatible change, because the old code path remains intact.

Updated:

* Responded more in-depth to review.
* Merged current master into this branch; manually resolved conflicts (that was required especially after https://github.com/python/cpython/pull/3058 and https://github.com/python/cpython/pull/1955 landed recently).
* Added code comments based on review feedback.
* Built successfully on Fedora 26's `gcc version 7.2.1` (new `dereferencing pointer to incomplete type` errors newly appeared compared to my work from a year ago).
* Confirmed that `./python -m test -v test_ssl` passes (one error message emitted by OpenSSL seems to have changed from OpenSSL 1.0.2 to 1.1.0, and I changed one `assertRaisesRegex(SSLError, ...)` back to simply `assertRaises(SSLError)`)
* Added NEWS entry.

jgehrcke committed:
Lib/ssl.py: fix indendation

jgehrcke committed:
bpo-16487: add entry to NEWS.d/next/Library

pitrou commented:
I'm still surprised that this functionality needs so much support code, but I'll let @tiran review it.

jgehrcke commented:
> I'm still surprised that this functionality needs so much support code, but I'll let @tiran review it.

Gotcha. At the core that's because there is unfortunately no such thing like `SSL_CTX_use_certificate_chain_bio()`  provided by the OpenSSL API: a `bio` pendant corresponding to `SSL_CTX_use_certificate_chain_file()` would be super convenient.

That's what this patch adds and what's called `_openssl_use_certificate_chain_from_bio()`.

Thanks for giving this another quick thought @pitrou, much appreciated.

**Edit:** I've tried to write this before, but I want to retry with more clarity: `_openssl_use_certificate_chain_from_bio()` is effectively `static int use_certificate_chain_file(SSL_CTX *ctx, SSL *ssl, const char *file)` straight from [the OpenSSL source](https://github.com/openssl/openssl/blob/d72a00416a0691bfd4920008767221bb4082a2ed/ssl/ssl_rsa.c#L589), just without BIO creation.

OpenSSL's `use_certificate_chain_file(SSL_CTX *ctx, SSL *ssl, const char *file)` expects a file path as input and then immediately performs BIO creation from a file (via `BIO_read_filename(in, file)`), and then processes this BIO with the *"OpenSSL cert chain loading logic"* (a term I will use again below).

In this patch, we create the BIO using the nice existing Python API for Memory BIO management (which was [added in Python 3.5](https://docs.python.org/3/library/ssl.html#memory-bio-support)). That results in `PySSLMemoryBIO` objects that hold the OpenSSL-native bio structs internally.

The input to the `_openssl_use_certificate_chain_from_bio()` added by this patch is precisely the native OpenSSL Bio struct. The function then proceeds with the  *"OpenSSL cert chain loading logic"* copied from OpenSSL' `use_certificate_chain_file(SSL_CTX *ctx, SSL *ssl, const char *file)`.

This is why `_openssl_use_certificate_chain_from_bio()` is OpenSSL's `use_certificate_chain_file(SSL_CTX *ctx, SSL *ssl, const char *file)` modulo BIO creation.

That
1) leaves the old `use_certificate_chain_file()`-based code path intact (for conceptually ensuring backwards compat).
2) should conceptually ensure that the new code path has the same effects on a given `SSLContext` when using the same input data, just from memory instead of from file.

tiran reviewed (changes requested):
I'm still -1 on this change. @reaperhulk and I are going to have another stab in PEP 543, which will provide a better way to load certs and keys. As Antoine pointed out, it's a lot of additional code. I don't want to maintain additional code and multiple APIs to load key and cert material.
